// @servoy/public/esm2015/lib/format/pipes.js
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/platform-browser";
export class MnemonicletterFilterPipe {
    transform(input, letter) {
        if (letter && input)
            return input.replace(letter, '<u>' + letter + '</u>');
        return input;
    }
}
MnemonicletterFilterPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: MnemonicletterFilterPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
MnemonicletterFilterPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: MnemonicletterFilterPipe, name: "mnemonicletterFilter" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: MnemonicletterFilterPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'mnemonicletterFilter' }]
        }] });
export class NotNullOrEmptyPipe {
    transform(value, _args) {
        if (value)
            return value.filter(a => a && (a.realValue || a.displayValue));
        return value;
    }
}
NotNullOrEmptyPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NotNullOrEmptyPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
NotNullOrEmptyPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NotNullOrEmptyPipe, name: "notNullOrEmpty" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NotNullOrEmptyPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'notNullOrEmpty' }]
        }] });
export class HtmlFilterPipe {
    transform(input) {
        if (input && input.indexOf && input.indexOf('<body') >= 0 && input.lastIndexOf('</body') >= 0) {
            input = input.substring(input.indexOf('<body') + 6, input.lastIndexOf('</body'));
        }
        return input;
    }
}
HtmlFilterPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: HtmlFilterPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
HtmlFilterPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: HtmlFilterPipe, name: "htmlFilter" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: HtmlFilterPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'htmlFilter' }]
        }] });
export class TrustAsHtmlPipe {
    constructor(domSanitizer) {
        this.domSanitizer = domSanitizer;
    }
    transform(input, trustAsHtml) {
        if (trustAsHtml) {
            return this.domSanitizer.bypassSecurityTrustHtml(input);
        }
        return input;
    }
}
TrustAsHtmlPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: TrustAsHtmlPipe, deps: [{ token: i1.DomSanitizer }], target: i0.ɵɵFactoryTarget.Pipe });
TrustAsHtmlPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: TrustAsHtmlPipe, name: "trustAsHtml" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: TrustAsHtmlPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'trustAsHtml' }]
        }], ctorParameters: function () { return [{ type: i1.DomSanitizer }]; } });
