// @servoy/public/esm2015/lib/directives/autosave.directive.js
import { Directive, HostListener } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../services/servoy_public.service";
export class AutosaveDirective {
    constructor(servoyService, elementRef) {
        this.servoyService = servoyService;
        this.elementRef = elementRef;
    }
    onClick(target) {
        if (target == this.elementRef.nativeElement || target.parentNode == this.elementRef.nativeElement) {
            this.servoyService.callService('applicationServerService', 'autosave', {}, true);
        }
    }
}
AutosaveDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: AutosaveDirective, deps: [{ token: i1.ServoyPublicService }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
AutosaveDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.2", type: AutosaveDirective, selector: "[svyAutosave]", host: { listeners: { "click": "onClick($event.target)" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: AutosaveDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[svyAutosave]'
                }]
        }], ctorParameters: function () { return [{ type: i1.ServoyPublicService }, { type: i0.ElementRef }]; }, propDecorators: { onClick: [{
                type: HostListener,
                args: ['click', ['$event.target']]
            }] } });
