// @servoy/public/esm2015/lib/directives/decimalkeyconverter.directive.js
import { Directive, Input, HostListener } from '@angular/core';
import { getLocaleNumberSymbol, NumberSymbol } from '@angular/common';
import * as i0 from "@angular/core";
import * as i1 from "../services/servoy_public.service";
export class DecimalkeyconverterDirective {
    constructor(el, servoyService) {
        this.el = el;
        this.servoyService = servoyService;
        this.element = el.nativeElement;
    }
    onKeypress(e) {
        if (e.which === 110 && this.svyFormat && this.svyFormat.type === 'NUMBER') {
            const caretPos = this.element.selectionStart;
            const startString = this.element.value.slice(0, caretPos);
            const endString = this.element.value.slice(this.element.selectionEnd, this.element.value.length);
            this.element.value = (startString + getLocaleNumberSymbol(this.servoyService.getLocale(), NumberSymbol.Decimal) + endString);
            this.element.focus();
            this.element.setSelectionRange(caretPos + 1, caretPos + 1);
            if (e.preventDefault)
                e.preventDefault();
        }
    }
}
DecimalkeyconverterDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: DecimalkeyconverterDirective, deps: [{ token: i0.ElementRef }, { token: i1.ServoyPublicService }], target: i0.ɵɵFactoryTarget.Directive });
DecimalkeyconverterDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.2", type: DecimalkeyconverterDirective, selector: "[svyDecimalKeyConverter]", inputs: { svyFormat: ["svyDecimalKeyConverter", "svyFormat"] }, host: { listeners: { "keydown": "onKeypress($event)" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: DecimalkeyconverterDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[svyDecimalKeyConverter]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.ServoyPublicService }]; }, propDecorators: { svyFormat: [{
                type: Input,
                args: ['svyDecimalKeyConverter']
            }], onKeypress: [{
                type: HostListener,
                args: ['keydown', ['$event']]
            }] } });
