// @servoy/public/esm2015/lib/directives/startedit.directive.js
import { Directive, Input, HostListener } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../logger.service";
export class StartEditDirective {
    constructor(logFactory) {
        this.log = logFactory.getLogger('StartEditDirective');
    }
    onFocus(e) {
        if (!this.hostComponent) {
            this.log.error('host component not found for the start edit directive use [hostComponent]="self" besides this in the template (component must be extending ServoyBaseComponent)');
        }
        else if (this.hostComponent.servoyApi && this.dataProviderID !== undefined) {
            this.hostComponent.servoyApi.startEdit(this.dataProviderID);
        }
        else {
            this.log.error('Can\'t call startEdit, missing servoyApi and dataProviderID for field ' + this.hostComponent);
        }
    }
}
StartEditDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: StartEditDirective, deps: [{ token: i1.LoggerFactory }], target: i0.ɵɵFactoryTarget.Directive });
StartEditDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.2", type: StartEditDirective, selector: "[svyStartEdit]", inputs: { dataProviderID: ["svyStartEdit", "dataProviderID"], hostComponent: "hostComponent" }, host: { listeners: { "focus": "onFocus($event)" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: StartEditDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[svyStartEdit]'
                }]
        }], ctorParameters: function () { return [{ type: i1.LoggerFactory }]; }, propDecorators: { dataProviderID: [{
                type: Input,
                args: ['svyStartEdit']
            }], hostComponent: [{
                type: Input
            }], onFocus: [{
                type: HostListener,
                args: ['focus', ['$event']]
            }] } });
