// @servoy/public/esm2015/lib/directives/upload.directive.js
import { Directive, Input, HostListener } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../services/servoy_public.service";
export class UploadDirective {
    constructor(servoyService) {
        this.servoyService = servoyService;
        this.propertyName = 'dataProviderID';
    }
    click(e) {
        this.servoyService.showFileOpenDialog('Please select a file', false, null, this.url);
    }
    ngOnInit() {
        this.url = this.servoyService.generateUploadUrl(this.formname, this.componentName, this.propertyName);
    }
}
UploadDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: UploadDirective, deps: [{ token: i1.ServoyPublicService }], target: i0.ɵɵFactoryTarget.Directive });
UploadDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.2", type: UploadDirective, selector: "[svyUpload]", inputs: { formname: "formname", componentName: "componentName" }, host: { listeners: { "click": "click()" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: UploadDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[svyUpload]'
                }]
        }], ctorParameters: function () { return [{ type: i1.ServoyPublicService }]; }, propDecorators: { formname: [{
                type: Input
            }], componentName: [{
                type: Input
            }], click: [{
                type: HostListener,
                args: ['click']
            }] } });
