// @servoy/public/esm2015/lib/tooltip/tooltip.directive.js
import { Directive, HostListener, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./tooltip.service";
export class TooltipDirective {
    constructor(tooltipService) {
        this.tooltipService = tooltipService;
        this.isActive = false;
        this.tooltipService.isTooltipActive.subscribe(a => {
            this.isActive = a;
        });
    }
    onMouseEnter(event) {
        if (this.tooltipText)
            this.tooltipService.showTooltip(event, this.tooltipText, 750, 5000);
    }
    onMouseLeave() {
        this.tooltipService.hideTooltip();
    }
    onClick() {
        this.tooltipService.hideTooltip();
    }
    onContextMenu() {
        this.tooltipService.hideTooltip();
    }
    ngOnDestroy() {
        this.tooltipService.hideTooltip();
    }
}
TooltipDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: TooltipDirective, deps: [{ token: i1.TooltipService }], target: i0.ɵɵFactoryTarget.Directive });
TooltipDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.2", type: TooltipDirective, selector: "[svyTooltip]", inputs: { tooltipText: ["svyTooltip", "tooltipText"] }, host: { listeners: { "mouseenter": "onMouseEnter()", "mouseleave": "onMouseLeave()", "click": "onClick()", "contextmenu": "onContextMenu()" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: TooltipDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[svyTooltip]' }]
        }], ctorParameters: function () { return [{ type: i1.TooltipService }]; }, propDecorators: { tooltipText: [{
                type: Input,
                args: ['svyTooltip']
            }], onMouseEnter: [{
                type: HostListener,
                args: ['mouseenter']
            }], onMouseLeave: [{
                type: HostListener,
                args: ['mouseleave']
            }], onClick: [{
                type: HostListener,
                args: ['click']
            }], onContextMenu: [{
                type: HostListener,
                args: ['contextmenu']
            }] } });
