// @servoy/public/esm2015/lib/servoy_public.module.js
import { NgModule } from '@angular/core';
import { TooltipDirective } from './tooltip/tooltip.directive';
import { TooltipService } from './tooltip/tooltip.service';
import { DecimalkeyconverterDirective } from './directives/decimalkeyconverter.directive';
import { MnemonicletterFilterPipe, NotNullOrEmptyPipe, HtmlFilterPipe, TrustAsHtmlPipe } from './format/pipes';
import { FormatFilterPipe } from './format/format.pipe';
import { EmptyValueFilterPipe } from './format/emptyvalue.pipe';
import { StartEditDirective } from './directives/startedit.directive';
import { ImageMediaIdDirective } from './directives/imagemediaid.directive';
import { AutosaveDirective } from './directives/autosave.directive';
import { UploadDirective } from './directives/upload.directive';
import { FormatDirective } from './format/formatcontrolvalueaccessor.directive';
import { FormattingService } from './format/formatting.service';
import { ComponentContributor } from './basecomponent';
import { SabloTabseq } from './directives/sablotabseq.directive';
import { WindowRefService } from './services/windowref.service';
import { SpecTypesService } from './spectypes.service';
import { LoggerFactory } from './logger.service';
import * as i0 from "@angular/core";
export class ServoyPublicModule {
}
ServoyPublicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: ServoyPublicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ServoyPublicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: ServoyPublicModule, declarations: [TooltipDirective,
        MnemonicletterFilterPipe,
        NotNullOrEmptyPipe,
        HtmlFilterPipe,
        FormatDirective,
        DecimalkeyconverterDirective,
        FormatFilterPipe,
        EmptyValueFilterPipe,
        StartEditDirective,
        ImageMediaIdDirective,
        AutosaveDirective,
        UploadDirective,
        SabloTabseq,
        TrustAsHtmlPipe], exports: [TooltipDirective,
        MnemonicletterFilterPipe,
        NotNullOrEmptyPipe,
        HtmlFilterPipe,
        FormatDirective,
        DecimalkeyconverterDirective,
        FormatFilterPipe,
        EmptyValueFilterPipe,
        StartEditDirective,
        ImageMediaIdDirective,
        AutosaveDirective,
        UploadDirective,
        SabloTabseq,
        TrustAsHtmlPipe] });
ServoyPublicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: ServoyPublicModule, providers: [TooltipService, FormattingService, ComponentContributor, SpecTypesService, WindowRefService, LoggerFactory], imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: ServoyPublicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TooltipDirective,
                        MnemonicletterFilterPipe,
                        NotNullOrEmptyPipe,
                        HtmlFilterPipe,
                        FormatDirective,
                        DecimalkeyconverterDirective,
                        FormatFilterPipe,
                        EmptyValueFilterPipe,
                        StartEditDirective,
                        ImageMediaIdDirective,
                        AutosaveDirective,
                        UploadDirective,
                        SabloTabseq,
                        TrustAsHtmlPipe
                    ],
                    imports: [],
                    exports: [TooltipDirective,
                        MnemonicletterFilterPipe,
                        NotNullOrEmptyPipe,
                        HtmlFilterPipe,
                        FormatDirective,
                        DecimalkeyconverterDirective,
                        FormatFilterPipe,
                        EmptyValueFilterPipe,
                        StartEditDirective,
                        ImageMediaIdDirective,
                        AutosaveDirective,
                        UploadDirective,
                        SabloTabseq,
                        TrustAsHtmlPipe
                    ],
                    providers: [TooltipService, FormattingService, ComponentContributor, SpecTypesService, WindowRefService, LoggerFactory]
                }]
        }] });
